### ISSUE TITLE
Global Interface Refactor/pimpl standardisation

### ISSUE THREAD

**Issue by Robadob** [FLAMEGPU/FLAMEGPU2#466]
As we near release, I'm growing more concerned over some of the more subtle interface inconsistencies. I wrote most of these interfaces, so i'm to blame for there not being a standard way of handling separation of interface and data.

## Example A

`AgentDescription/AgentData`: (also applicable to most of the `ModelDescription` hierarchy)

`ModelDescription::newAgent/Agent/getAgent()` returns `AgentDescription&`, a reference to an interface owned by it's `AgentData`.

This means a user cannot copy the interface, whether by accident or intentionally, copy and assignment operators are explicitly deleted.

The need for a user to specify `AgentDescription&` (but not in other places could cause confusion), it's possible for them to take the `ModelDescription` out of scope, and the reference point to something undefined..

## Example B 

`EnvironmentDescription`

`ModelDescription::Environment/getEnvironmen()` returns `EnvironmentDescription&`, a reference to an object owned by it's `AgentData`.

`EnvironmentDescription` is not an interface like it's other counterparts, it is both an interface and holds the data.
`EnvironmentDescription` hasn't got deleted copy/assignment operators, so a user can take a copy and edit that with  no impact to anything. This is bad.

The need for a user to specify `EnvironmentDescription&`, same as above.

## Example C

`HostAgentAPI`

`HostAPI::agent/newAgent()` returns `HostAgentAPI`, a unique instance of an interface to an underlying shared data structure

As `HostAgentAPI` is merely a shell interface, it can be copied/assigned at will, and will point to the data of the provided `HostAgentAPI`.

Example C is the most transparent and flexible of the 3.

-------------

# How to Address this?

## Short-term

1.

Any interfaces which are returned by reference should be renamed in the form `AgentDescription` -> `AgentDescription_t`.
A typedef should then be declared in their place, `typedef AgentDescription_t& AgentDescription`.

(This is similar to how some libs, e.g. CUDA hide pointers)

Now a user does not need to specify that they are receiving a reference, it is hidden to them, all interfaces look the same.

2. 
`EnvironmentDescription` should have it's copy/assignment operators marked as `deleted`.

Now a user can't accidentally make a copy.

3.

Detail a list of all the 'interfaces' exposed to a user, and group them according to their status (there might be a 4th state not included in the list above e.g. `DeviceAPI`).

## Long-term

If the short-term changes are applied, these long-term changes should not change the interface, so they will be invisible to users.

1. Split `EnvironmentDescription` into `EnvironmentDescription` and `EnvironmentData`.

2. Update the entire `ModelDescription` hierarchy, so that Description classes are standalone instances which merely hold a weak_ptr, to their related data class. As such they can be copied/assigned as desired.

3. Find a solution of `DeviceAgentVector`, it inherits from `AgentVector`. It probably needs an interface class adding, it will simply be that the current `DeviceAgentVector` remains a complicated class,, rather than a more basic data struct.

4. Address any other interfaces returned by HostAPI to match the entire `ModelDescription` hierarchy style.

5. Add copy and assignment tests to all interface classses.

Worth giving this a read before attempt to address this: https://en.cppreference.com/w/cpp/language/pimpl

**Comment by Robadob** [comment]
Noticed a slight issue with short-term solution 1.

In the below case.

```c

typedef foo& bar;

bar foobar() {
  // return something
}

auto t = foobar();
```

`auto` resolves to `foo` rather than `foo&`. In the case of the current `DeviceAgentVector` implementation this creates an awful compile error where it tries to perform copy construction and gets blocked by some template failure.
